Guide/src/dev_guide/dev_tools/xflowey.md

```diff
@@ -10,13 +10,885 @@ for writing maintainable, cross-platform automation.
+);
+```
+
+### StepCtx vs NodeCtx
+
+- **`NodeCtx`**: Used when emitting steps. Provides `emit_*` methods, `new_var()`,
+  `req()`, etc.
+  
+- **`StepCtx`**: Used inside step closures. Provides access to `claim()` for
+  variables, and basic environment info (`backend()`, `platform()`).
+
+---
+
+## Runtime Services
+
+Runtime services provide the API available during step execution (inside the
+closures passed to `emit_rust_step`, etc.).
+
+### RustRuntimeServices
+
+Available in Rust steps via the `rt` parameter:
+
+```rust
+move |rt: &mut RustRuntimeServices<'_>| {
+    // Read variables
+    let value = rt.read(some_var);
+    
+    // Write variables
+    rt.write(output_var, &result);
+    rt.write_secret(secret_var, &secret);  // Mark as secret
+    
+    // Query environment
+    let backend = rt.backend();    // Local, ADO, or Github
+    let platform = rt.platform();  // Windows, Linux, MacOs
+    let arch = rt.arch();          // X86_64, Aarch64
+    
+    Ok(())
+}
+```
+
+**Important**: If a step reads a secret value, all subsequent writes from that
+step are marked as secret by default (to prevent accidental leaks). Use
+`write_not_secret()` if you need to override this.
+
+### AdoStepServices
+
+Available in ADO steps for interacting with ADO-specific features:
+
+```rust
+move |rt: &mut AdoStepServices<'_>| {
+    // Get ADO variable as flowey var
+    rt.set_var(flowey_var, AdoRuntimeVar::BUILD__SOURCE_BRANCH);
+    
+    // Set ADO variable from flowey var
+    let ado_var = rt.get_var(flowey_var);
+    
+    // Resolve repository ID
+    let repo = rt.resolve_repository_id(repo_id);
+    
+    "- task: SomeTask@1".to_string()
+}
+```
+
+### GhStepBuilder
+
+Builder for GitHub Actions steps:
+
+```rust
+ctx.emit_gh_step("Azure Login", "Azure/login@v2")
+    .with("client-id", client_id)           // Add parameter
+    .with("tenant-id", tenant_id)
+    .output("token", token_var)             // Capture output
+    .run_after(some_side_effect)            // Add dependency
+    .requires_permission(                   // Declare permission needed
+        GhPermission::IdToken,
+        GhPermissionValue::Write
+    )
+    .finish(ctx);
+```
+
+---
+
+## Node Design Philosophy
+
+Flowey nodes are designed around several key principles:
+
+### 1. Composability
+
+Nodes should be reusable building blocks that can be combined to build complex
+workflows. Each node should have a single, well-defined responsibility.
+
+❌ **Bad**: A node that "builds and tests the project"  
+✅ **Good**: Separate nodes for "build project" and "run tests"
+
+### 2. Explicit Dependencies
+
+Dependencies between steps should be explicit through variables, not implicit
+through side effects.
+
+❌ **Bad**: Assuming a tool is already installed  
+✅ **Good**: Taking a `ReadVar<SideEffect>` that proves installation happened
+
+### 3. Backend Abstraction
+
+Nodes should work across all backends when possible. Backend-specific behavior
```

Comment:
Is it preferred to `unimplemented!` or `anyhow::bail!` if a node is executed with a backend it doesn't expect? E.g. if a node is _only_ viable in ADO, what should it do if the backend is Local or GH?